psycopg/psycopg/types/geometry.py

```diff
@@ -0,0 +1,86 @@
+"""
+Adapters for PostGIS geometries
+"""
+
+from typing import Optional, Type
+
+from .. import postgres
+from ..abc import AdaptContext
+from ..adapt import Dumper, Loader
+from ..pq import Format
+from .._typeinfo import TypeInfo
+
+
+try:
+    import shapely.wkb as wkb
+    from shapely.geometry.base import BaseGeometry
+
+except ImportError:
+    raise ImportError(
+        "The module psycopg.types.geometry requires the package 'Shapely'"
+        " to be installed"
+    )
+
+
+class GeometryBinaryLoader(Loader):
+    format = Format.BINARY
+
+    def load(self, data: bytes) -> "BaseGeometry":
+        return wkb.loads(data)
+
+
+class GeometryLoader(Loader):
+    format = Format.TEXT
+
+    def load(self, data: bytes) -> "BaseGeometry":
+        # it's a hex string in binary
+        return wkb.loads(data.decode(), hex=True)
+
+
+class GeometryBinaryDumper(Dumper):
+    format = Format.BINARY
+
+    def dump(self, obj: "BaseGeometry") -> bytes:
+        return wkb.dumps(obj).encode()  # type: ignore
+
+
+class GeometryDumper(Dumper):
+    format = Format.TEXT
+
+    def dump(self, obj: "BaseGeometry") -> bytes:
+        return wkb.dumps(obj, hex=True).encode()  # type: ignore
+
+
+def register_shapely(
+    info: TypeInfo, context: Optional[AdaptContext] = None
+) -> None:
+    """Register Shapely dumper and loaders.
+
+    After invoking this function on an adapter, the queries retrieving
+    PostGIS geometry objects will return Shapely's shape object instances
+    both in text and binary mode.
+
+    Similarly, shape objects can be sent to the database.
+
+    This requires the Shapely library to be installed.
+
+    :param info: The object with the information about the geometry type.
+    :param context: The context where to register the adapters. If `!None`,
+        register it globally.
+
+    """
+
+    info.register(context)
+    adapters = context.adapters if context else postgres.adapters
+    # Generate and register the text and binary dumper
+    binary_dumper: Type[GeometryBinaryDumper] = type(
+        "GeometryBinaryDumper", (GeometryBinaryDumper,), {"oid": info.oid}
+    )
+    dumper: Type[GeometryDumper] = type(
+        "GeometryDumper", (GeometryDumper,), {"oid": info.oid}
+    )
+
+    adapters.register_loader(info.oid, GeometryBinaryLoader)
+    adapters.register_loader(info.oid, GeometryLoader)
+    adapters.register_dumper(BaseGeometry, binary_dumper)
+    adapters.register_dumper(BaseGeometry, dumper)
```

Comment:
Registering dumpers on the same type, the last one becomes the default dumper. The way you registered them here make the text dumper the default one. I assume you would prefer the binary one to be the default, right? If that's the case you may want to swap these two `register_dumper()`

Reply:
Isn't it also stored based on the format ([here](https://github.com/psycopg/psycopg/blob/master/psycopg/psycopg/_adapters_map.py#L142)) ? I'm not sure what that `type` wizardry does here exactly but I expect the `format` attribute to be defined by the two different base classes so that the corresponding one is used when a cursor is created with `binary=True` or not

Reply:
`type()` is just a way to create a class using a function instead of using the `class` syntax. In your case tbh the class would have worked as well (in other places I think I have, or had, definitions in a loop or something like that). Changing the base class name to disambiguate, the call:
```python
dumper = type("GeometryDumper", (BaseGeometryDumper,), {"oid": info.oid})
```
is equivalent to:
```python
class GeometryDumper(BaseGeometryDumper):
    oid = info.oid

dumper = GeometryDumper
```

About the formats, psycopg behaviour is [explained here](https://www.psycopg.org/psycopg3/docs/basic/params.html#binary-data). When passing parameters to the client, each value can be sent as text or binary, independently. When the `%s` placeholder is used, psycopg uses the most convenient format: usually but not always binary. In order to force the text or binary format in a query, the user can use `%t` or `%b` placeholders.

On the way back, postgres will return all the data either as text or as binary: that's what `binary=True` on the cursor or query controls.

So, Dumpers can be text or binary, yes, but there is also the choices of "which one to use by default when the user doesn't care?" I think in this case the binary one works better, because it's complete and more efficient. Which one is the default one? The last one registered.

https://github.com/psycopg/psycopg/blob/b70e9da3bedc408e4e4f88ab68b892f5d4d0fd25/psycopg/psycopg/_adapters_map.py#L134-L135

Reply:
I see, it makes sense now.